File: recursividade/javascript/palindromo.js
function isPalindrome(text) {
  return isPalindromeTailRecursive(text, 0, text.length - 1);
}

function isPalindromeTailRecursive(text, start, end) {
  if (start >= end) {
    return true;
  }

  if (text[start] !== text[end]) {
    return false;
  }

  return isPalindromeTailRecursive(text, start + 1, end - 1);
}

console.log(isPalindrome(""));
console.log(isPalindrome("aba"));
console.log(isPalindrome("abccba"));
console.log(isPalindrome("abcfba"));

File: recursividade/java/src/Palindromo.java
public class Palindromo {
	public static boolean isPalindrome(String text) {
        return isPalindromeTailRecursive(text, 0, text.length() - 1);
    }

    private static boolean isPalindromeTailRecursive(String text, int start, int end) {
        if (start >= end) {
            return true;
        }

        if (text.charAt(start) != text.charAt(end)) {
            return false;
        }

        return isPalindromeTailRecursive(text, start + 1, end - 1);
    }

    public static void main(String[] args) {
        System.out.println(isPalindrome(""));
        System.out.println(isPalindrome("aba"));
        System.out.println(isPalindrome("abccba"));
        System.out.println(isPalindrome("abcfba"));
    }
}
